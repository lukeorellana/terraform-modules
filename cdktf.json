{
  "language": "typescript",
  "app": "npx ts-node",
  "terraformProviders": [],
  "projectId": "1b5f16bd-67ac-4af6-a33f-7bcafff141eb"
}