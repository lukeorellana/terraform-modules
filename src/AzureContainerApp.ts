import { Fn } from 'cdktf';
import { Construct } from 'constructs';
import * as azapi from "../.gen/providers/azapi"

// Construct
/**
 * Properties for the resource group
 */

 export interface AzureContainerAppsProps {
  /**
   * The Azure Region to deploy.
   */
  readonly location: string;
  /**
   * The name of the Azure Container App.
   */
  readonly name: string;
  /**
   * The id of the Azure Resource Group.
   */
   readonly resource_group_id: string;
  /**
   * The tags to assign to the Resource Group.
   */
   readonly tags?: { [key: string]: string; };  
   /**
   * The container name.
   */
   readonly container_name: string;
   /**
   * The container port.
   */
   readonly container_port: number;
   /**
   * The CPU requests to run each container.
   */
   readonly cpu_requests: number;
   /**
   * The container image to deploy.
   */
    readonly image: string;
    /**
   * Enabled ingress.
   */
    readonly ingress_enabled: boolean;
    /**
   * Max amount of replica's container app can scale to.
   */
    readonly max_replicas: number;
    /**
   * Memory request of container app.
   */
    readonly mem_requests: string;
    /**
   * Minimum amount of replicas to kee.
   */
    readonly min_replicas: number;
  /**
   * The container tag of the image to deploy.
   */
    readonly container_tag: string;
  /**
   * The container tag of the image to deploy.
   */
    readonly loganalytics_id: string
      /**
   * The container tag of the image to deploy.
   */
    readonly loganalytics_primary_key: string
    /**
   * The hostname of the Azure container registry.
   */
     readonly acr_url: string
   
}

export class AzureContainerApp extends Construct {
  readonly props: AzureContainerAppsProps;
  

  constructor(scope: Construct, id: string, props: AzureContainerAppsProps) {
    super(scope, id);

    this.props = props;;

    const azapiEnvironment = new azapi.Resource(this, "managed_environment", {
      body:  Fn.jsonencode({ 
        properties: { 
          appLogsConfiguration: {
            destination: "log-analytics", 
            logAnalyticsConfiguration: {
              customerId: props.loganalytics_id,
              sharedKey:   props.loganalytics_primary_key,
          }
        }
      }
    }),
      location: props.location,
      name: props.name + "-env",
      parentId: props.resource_group_id,
      tags: props.tags,
      type: "Microsoft.App/managedEnvironments@2022-03-01",
    });

    

   const aca = new azapi.Resource(this, "container_app", {
      body: Fn.jsonencode({ 
        "identity": {
            "type": "SystemAssigned"
        },
        properties: {
          managedEnvironmentId: azapiEnvironment.id,
          configuration: {
            "registries": [
              {
                  "server": props.acr_url,
                  "identity": "system"
              }
            ],
            ingress: {
              external: props.ingress_enabled,
              targetPort: props.container_port
            }
          },
          template: {
            containers: [
              {
                name: props.container_name,
                image: props.image + ":" + props.container_tag,
                resources: {
                  cpu: props.cpu_requests,
                  memory: props.mem_requests
                }
              }         
            ],
            scale: {
              minReplicas: props.min_replicas,
              maxReplicas: props.max_replicas
            }
          }
        }


      }),
      location: props.location,
      name: props.name,
      parentId: props.resource_group_id,
      tags: props.tags,
      type: "Microsoft.App/containerApps@2022-03-01",
      
    });


    azapiEnvironment.overrideLogicalId("managed_environment")
    aca.overrideLogicalId("container_app")
  }

}
