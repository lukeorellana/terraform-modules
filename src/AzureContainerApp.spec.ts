import { Testing, TerraformStack, testingMatchers} from 'cdktf';
import {AzurermProvider} from "@cdktf/provider-azurerm";
import { AzureContainerApp } from './AzureContainerApp';
import {AzapiProvider} from "../.gen/providers/azapi"
import 'cdktf/lib/testing/adapters/jest';


// test('renders a VPC with minimal config', () => {
//   const synthed = Testing.synthScope((stack) => {
//     new PocketVPC(stack, 'testPocketVPC');
//   });


describe('AzureContainerApp-Snapshot', () => {
  it('renders a AzureContainerApp', () => {

    const synthed = Testing.synthScope((stack) => {
      new AzureContainerApp(stack, 'testACA', {
        name: 'acrtest888',
        location: 'eastus',
        resource_group_id: "rg-test",
        container_name: "test-container",
        container_port: 8080,
        cpu_requests: 2,
        image: "uzyexe/tetris",
        ingress_enabled: true,
        max_replicas: 3,
        mem_requests: "1Gi",
        min_replicas: 1,
        container_tag: "latest",
        loganalytics_id: "loganalyticsid",
        loganalytics_primary_key: "loganalyticskey",
        acr_url: "testacr.azurecr.io"
      });
    });
  
    expect(synthed).toMatchSnapshot();
  });
});


describe("AzureContainerApps-Terraform", () => {
  it("check if the produced terraform configuration is valid", () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "test");
    
    new AzurermProvider(stack, "azureFeature", {
      features: {},
    });

    new AzapiProvider(stack, "azureAzapi", {
    });


    new AzureContainerApp(stack, 'testACA', {
      name: 'acrtest888',
      location: 'eastus',
      resource_group_id: "rg-test",
      container_name: "test-container",
      container_port: 8080,
      cpu_requests: 2,
      image: "uzyexe/tetris",
      ingress_enabled: true,
      max_replicas: 3,
      mem_requests: "1Gi",
      min_replicas: 1,
      container_tag: "latest",
      loganalytics_id: "loganalyticsid",
      loganalytics_primary_key: "loganalyticskey",
      acr_url: "testacr.azurecr.io"
    });


    // We need to do a full synth to validate the terraform configuration
    expect(Testing.fullSynth(stack)).toBeValidTerraform();
  });

  it("check if this can be planned", () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureFeature", {
      features: {},
    });

    new AzapiProvider(stack, "azureAzapi", {
    });

    new AzureContainerApp(stack, 'testACA', {
      name: 'acrtest888',
      location: 'eastus',
      resource_group_id: "rg-test",
      container_name: "test-container",
      container_port: 8080,
      cpu_requests: 2,
      image: "uzyexe/tetris",
      ingress_enabled: true,
      max_replicas: 3,
      mem_requests: "1Gi",
      min_replicas: 1,
      container_tag: "latest",
      loganalytics_id: "loganalyticsid",
      loganalytics_primary_key: "loganalyticskey",
      acr_url: "testacr.azurecr.io"
    });


    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(stack)).toPlanSuccessfully();
  });
});
