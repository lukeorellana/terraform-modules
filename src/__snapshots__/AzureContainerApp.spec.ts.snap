// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AzureContainerApp-Snapshot renders a AzureContainerApp 1`] = `
"{
  "resource": {
    "azapi_resource": {
      "testACA_aca_15FAF6B7": {
        "body": "\${jsonencode({\\n    properties: {\\n      managedEnvironmentId = \${azapi_resource.testACA_aca_env_CF086A3B.id}\\n      configuration = {\\n        ingress = {\\n          external = true\\n          targetPort = true?8080: null\\n        }\\n      }\\n      template = {\\n        containers = [\\n          {\\n            name = \\"main\\"\\n            image = \\"uzyexe/tetris:latest\\"\\n            resources = {\\n              cpu = 2\\n              memory = 1.0Gi\\n            }\\n          }         \\n        ]\\n        scale = {\\n          minReplicas = 1\\n          maxReplicas = 3\\n        }\\n      }\\n    }\\n  })}",
        "location": "eastus",
        "name": "acrtest888",
        "parent_id": "rg-test",
        "type": "Microsoft.App/containerApps@2022-03-01"
      },
      "testACA_aca_env_CF086A3B": {
        "body": "\${jsonencode({\\n    properties = {\\n      appLogsConfiguration = {\\n        destination               = \\"log-analytics\\"\\n        logAnalyticsConfiguration = {\\n          customerId = loganalyticsid\\n          sharedKey  = loganalyticskey\\n        }\\n      }\\n    }\\n })}",
        "location": "eastus",
        "name": "\${props.name}-env",
        "parent_id": "rg-test",
        "type": "Microsoft.App/managedEnvironments@2022-03-01"
      }
    }
  }
}"
`;
